pkg/services/publicdashboards/api/api.go

```diff
@@ -104,24 +109,66 @@ func (api *Api) GetPublicDashboard(c *models.ReqContext) response.Response {
+		return api.handleError(c.Req.Context(), http.StatusInternalServerError, "GetPublicDashboard: failed to get public dashboard ", err)
+	}
+
+	if pdc == nil {
+		return api.handleError(c.Req.Context(), http.StatusNotFound, "GetPublicDashboard: public dashboard not found", ErrPublicDashboardNotFound)
 	}
+
 	return response.JSON(http.StatusOK, pdc)
 }
 
-// SavePublicDashboard Sets public dashboard configuration for dashboard
-// POST /api/dashboards/uid/:uid/public-config
-func (api *Api) SavePublicDashboard(c *models.ReqContext) response.Response {
+// SavePublicDashboard Sets public dashboard for dashboard
+// POST /api/dashboards/uid/:uid/public-dashboards
+func (api *Api) CreatePublicDashboard(c *models.ReqContext) response.Response {
+	// exit if we don't have a valid dashboardUid
+	dashboardUid := web.Params(c.Req)[":dashboardUid"]
+	if dashboardUid == "" || !util.IsValidShortUID(dashboardUid) {
+		return api.handleError(c.Req.Context(), http.StatusBadRequest, "SavePublicDashboard: invalid dashboardUid", dashboards.ErrDashboardIdentifierInvalid)
+	}
+
+	pubdash := &PublicDashboard{}
+	if err := web.Bind(c.Req, pubdash); err != nil {
+		return api.handleError(c.Req.Context(), http.StatusBadRequest, "CreatePublicDashboard: bad request data", err)
+	}
+
+	// Always set the orgID and userID from the session
+	pubdash.OrgId = c.OrgID
+	dto := SavePublicDashboardDTO{
+		UserId:          c.UserID,
+		OrgId:           c.OrgID,
+		DashboardUid:    dashboardUid,
+		PublicDashboard: pubdash,
+	}
+
+	//Create the public dashboard
+	pubdash, err := api.PublicDashboardService.Create(c.Req.Context(), c.SignedInUser, &dto)
+	if err != nil {
+		return api.handleError(c.Req.Context(), http.StatusInternalServerError, "CreatePublicDashboard: failed to save public dashboard", err)
+	}
+
+	return response.JSON(http.StatusOK, pubdash)
+}
+
+// UpdatePublicDashboard Sets public dashboard for dashboard
+// PUT /api/dashboards/uid/:uid/public-dashboards
+func (api *Api) UpdatePublicDashboard(c *models.ReqContext) response.Response {
 	// exit if we don't have a valid dashboardUid
 	dashboardUid := web.Params(c.Req)[":dashboardUid"]
 	if dashboardUid == "" || !util.IsValidShortUID(dashboardUid) {
-		api.handleError(c.Req.Context(), http.StatusBadRequest, "SavePublicDashboard: invalid dashboardUid", dashboards.ErrDashboardIdentifierNotSet)
+		return api.handleError(c.Req.Context(), http.StatusBadRequest, "UpdatePublicDashboard: no dashboardUid", dashboards.ErrDashboardIdentifierInvalid)
 	}
 
 	pubdash := &PublicDashboard{}
 	if err := web.Bind(c.Req, pubdash); err != nil {
-		return response.Error(http.StatusBadRequest, "SavePublicDashboard: bad request data", err)
+		return api.handleError(c.Req.Context(), http.StatusBadRequest, "UpdatePublicDashboard: bad request data", err)
+	}
+
+	if pubdash.Uid == "" {
+		return api.handleError(c.Req.Context(), http.StatusBadRequest, "UpdatePublicDashboard: missing public dashboard uid", ErrPublicDashboardIdentifierNotSet)
```

Comment:
is this necessary? Take into account we're getting the pubdash in the service in this way 
```
existingPubdash, err := pd.store.Find(ctx, dto.PublicDashboard.Uid)
	if err != nil {
		return nil, err
	} else if existingPubdash == nil {
		return nil, ErrPublicDashboardNotFound
	}
```
so, if Uid is nil, we'll get an error. Otherwise, we should check for every attribute in the dto. I think it's an extra check that doesn't worth

Reply:
I think the `uid` should be a URL param because it is the identifier of the resource that we are updating. We should validate all attributes that are required and to be sure the resource is not going to break (e.g. empty accessToken). What happens if we receive an empty `accessToken` or empty `dashboardUid`? Should we allow updating all attributes?

Reply:
Thanks for bringing this up- I briefly thought about updating this and changed my approach a couple of times during the refactor.

The end goal is a proper api, which means that if we're updating a resource, we should reference it by it's uid. It's an easy change now and should we support multiple public dashboards in the future, we don't need to make any api changes.